### PR TITLE
rpcwallet: Add missing transaction categories to rpc helptexts

### PR DESCRIPTION
**Pull request by andrewtoth**
The current helptext for `listtransactions`, `listsinceblock` and `gettransaction` only list two of the five possible options for `category`. This incorrectly implies that these are the only two options, and can cause problems if the other three options aren't accounted for. Also, some of the documentation is incorrect when specifying which options are returned for which categories.

This PR updates the helptext for these RPCs and adds a functional regression test for the cases when the other three categories are returned.

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -1346,7 +1346,7 @@ UniValue listtransactions(const JSONRPCRequest& request)
             "[\n"
             "  {\n"
             "    \"address\":\"address\",    (string) The bitcoin address of the transaction.\n"
-            "    \"category\":\"send|receive\", (string) The transaction category.\n"
+            "    \"category\":\"send|receive|generate|immature|orphan\", (string) The transaction category.\n"
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Could improve test/functional/wallet_listtransactions.py with these categories.

Comment (andrewtoth) [reply src/wallet/rpcwallet.cpp]: Done

**Comment by MeshCollider** [comment]
Thanks, utACK https://github.com/bitcoin/bitcoin/pull/14653/commits/31066f523fc11ef8ba7943693180e56e04fd0bac
The `listsinceblock` RPC helptext also still refers to the 'move' category which was removed when accounts were, you could fix that up too in the same commit if you want :)

**Comment by andrewtoth** [comment]
@MeshCollider done. Also updated the `gettransaction` help.

**Comment by MeshCollider** [comment]
LGTM, re-utACK https://github.com/bitcoin/bitcoin/pull/14653/commits/4a78c4888a63bd2858b72744cf09e0dd7687c4be

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -1346,9 +1346,13 @@ UniValue listtransactions(const JSONRPCRequest& request)
             "[\n"
             "  {\n"
             "    \"address\":\"address\",    (string) The bitcoin address of the transaction.\n"
-            "    \"category\":\"send|receive\", (string) The transaction category.\n"
-            "    \"amount\": x.xxx,          (numeric) The amount in " + CURRENCY_UNIT + ". This is negative for the 'send' category, and is positive\n"
-            "                                        for the 'receive' category,\n"
+            "    \"category\":\"send|receive|generate|immature|orphan\",     (string) The transaction category.\n"
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: You could improve formatting by, for instance, following:
https://github.com/bitcoin/bitcoin/blob/ae32806ea20235c4542b1a5d4011df5ae267b58a/src/wallet/rpcwallet.cpp#L3017-L3023

**Comment by chris-belcher** [comment]
tACK `a2c555d`

**Comment by andrewtoth** [comment]
@promag Are the formatting and tests ok?

@MeshCollider Sorry to keep invalidating your utACKs. I updated to include category tests for all 3 rpcs.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -1363,9 +1363,13 @@ UniValue listtransactions(const JSONRPCRequest& request)
             "[\n"
             "  {\n"
             "    \"address\":\"address\",    (string) The bitcoin address of the transaction.\n"
-            "    \"category\":\"send|receive\", (string) The transaction category.\n"
-            "    \"amount\": x.xxx,          (numeric) The amount in " + CURRENCY_UNIT + ". This is negative for the 'send' category, and is positive\n"
-            "                                        for the 'receive' category,\n"
+            "    \"category\":               (string) The transaction category.\n"
+            "                \"send\"|                 'send' specifies transactions sent.\n"
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: I don't see the need to repeat. Something along this?
 ```
 "    \"category\"                    (string) The transaction category. One of\n" 
 "        \"send\"                    transactions sent\n"
 "        \"receive\"                 non-coinbase transations received\n" 
...
```
However I don't have strong opinion here and currently there is no "standard" output defined (that I know of).

Comment (andrewtoth) [reply src/wallet/rpcwallet.cpp]: Reformatted to remove the repetition.

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -1369,9 +1369,14 @@ UniValue listtransactions(const JSONRPCRequest& request)
             "[\n"
             "  {\n"
             "    \"address\":\"address\",    (string) The bitcoin address of the transaction.\n"
-            "    \"category\":\"send|receive\", (string) The transaction category.\n"
+            "    \"category\":               (string) The transaction category.\n"
+            "                \"send\"                  Transactions sent.\n"
+            "                \"receive\"               Non-coinbase transations received.\n"
```
Comment (practicalswift) [inline src/wallet/rpcwallet.cpp]: Should be "transactions" :-)

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -1496,20 +1501,25 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
...
             "{\n"
             "  \"transactions\": [\n"
-            "    \"address\":\"address\",    (string) The bitcoin address of the transaction. Not present for move transactions (category = move).\n"
-            "    \"category\":\"send|receive\",     (string) The transaction category. 'send' has negative amounts, 'receive' has positive amounts.\n"
-            "    \"amount\": x.xxx,          (numeric) The amount in " + CURRENCY_UNIT + ". This is negative for the 'send' category, and for the 'move' category for moves \n"
-            "                                          outbound. It is positive for the 'receive' category, and for the 'move' category for inbound funds.\n"
+            "    \"address\":\"address\",    (string) The bitcoin address of the transaction..\n"
+            "    \"category\":               (string) The transaction category.\n"
+            "                \"send\"                  Transactions sent.\n"
+            "                \"receive\"               Non-coinbase transations received.\n"
```
Comment (practicalswift) [inline src/wallet/rpcwallet.cpp]: Same typo here.

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -1645,7 +1655,12 @@ static UniValue gettransaction(const JSONRPCRequest& request)
             "  \"details\" : [\n"
             "    {\n"
             "      \"address\" : \"address\",          (string) The bitcoin address involved in the transaction\n"
-            "      \"category\" : \"send|receive\",    (string) The category, either 'send' or 'receive'\n"
+            "      \"category\" :                      (string) The transaction category.\n"
+            "                   \"send\"                  Transactions sent.\n"
+            "                   \"receive\"               Non-coinbase transations received.\n"
```
Comment (practicalswift) [inline src/wallet/rpcwallet.cpp]: And here :-)

Comment (andrewtoth) [reply src/wallet/rpcwallet.cpp]: All done

**Review by ryanofsky — APPROVED** [review]
utACK f3f6dde56ead45072864123ed0a2bb04ec584b65. Clearly a good change. It cleans up inaccurate documentation and adds a test.

**Comment by ryanofsky** [comment]
This looks like it could be merged.

**Comment by promag** [comment]
utACK f3f6dde.

**Comment by luke-jr** [comment]
utACK

**Pull request merged**